### ISSUE TITLE
File > Duplicate does not copy Additional Content into new document package

### ISSUE THREAD

**Issue by jerrykrinock** [karelia/BSManagedDocument#45]
When you duplicate a new document by clicking in the menu File > Duplicate, the newly-created document package does not contain copies of the so-called _additional content_ files from the original document package.  It seems to me that it should; at least, I need that in an app I'm now working on.  Does karelia agree?  Would karelia want a pull request if I could fix this?

Just to be clear, here are the exact Steps To Reproduce.
1. Build and launch the BSTest application.
2. Save the new document which opens as Document1.bstest.
3. Quit BSTest.
4. Run Finder.
5. Navigate to Document1.bstest and _Show Package Contents_.
6. Create a text file, test.txt, and save it inside the Document1.bstest package.
7. Relaunch BSTest.
8. Reopen Document1.bstest.
9. Click in the menu: File > Duplicate.  A new document appears.
10. Click in the menu: File > Save.
11. Save this new docu,ent as Document2.bstest.
12. Run Finder.
13. Navigate to Document2.bstest and "Show Package Contents".

Expected Result:  Document2.btest package should contain a copy of test.txt.
Actual Result: Document2.btest package contains only StoreContent.

**Comment by mikeabdullah** [comment]
Hi Jerry, do I understand you're talking in step 6 about modifying the package manually, rather than having code in your document subclass that does it?

**Comment by jerrykrinock** [comment]
Yes, that's how I tested it with the demo app to quickly and simply prove the point here.  I didn't think it would make any difference.  In the real app I'm developing, I do it the correct way, overriding -writeAdditionalContent::::.  But the real app has the same issue.

**Comment by mikeabdullah** [comment]
It should make a difference which way you go about it. Out of the box, I've taken the same approach with `BSManagedDocument` as with `NSDocument`. When saving to a new location, the default is to have the document _write_ itself to a new location. That means the document decides exactly what content should be present (generally this decision is made based on in-memory state) and writes it all to the new location.

Your scenario above would only work if the approach was to instead start by copying the existing files on disk, and then performing any further modifications that might be required. You could potentially have your document subclass start working in this fashion if it suits your needs better, although might be tricky.

One good reason for Apple/my approach is handling deleting of files from the document. If you go on the basis of copying the document around, and then modifying it, that means you have to implement some sort of extra change tracking of the files that need to be deleted/skipped. By _writing_ the doc effectively afresh each time, there's no need for such tracking since the new version of the doc replaces the old each time.

---

All the above said, you _should_ find that `writeAdditionalContent…` is called during a duplication. If it's not, are you able to figure out why not, or extend the sample project to show that?

**Comment by jerrykrinock** [comment]
Thank you, Mike.  After spending an afternoon going down the NSDocument Rabbit Hole, I have some conclusions but would rather wait until I can see daylight before discussing further.  I shall be back.

**Comment by jerrykrinock** [comment]
Thank you very much for the explanation, Mike.

I think that `writeAdditionalContent::::` is _not_ called during a _duplicate_ operation as you expected.  I mean, it is not called when user runs BSTest and clicks File > Duplicate.  You suggested that I determine _why not_.  Here you go…

There are only two calls to `writeAdditionalContent::::`, and both are in the big _worker block_ (lines 352-530) which is created and returned by `contentsForURL::::` and invoked later by `writeToURL:`.  During a _save_ operation, `contentsForURL::::` is called to create this block, but during a _duplicate_ operation, it is not.  So, the question becomes “why is `contentsForURL::::` not being called?”

Well, there are two calls to `contentsForURL::::`.  One of them (line 592) is in `saveToURL::::`, and indeed this is the call that does the job during a _save_ operation.  The other (line 781) is in `writeToURL:::::`, and this is the one that I think should occur during a _duplicate_ operation.  However it does not, because it is enclosed by `if (!_contents) {…}` (line 779), and `_contents` has already had another block assigned to it, at the beginning of the _duplicate_ operation, in `duplicateAndReturnError:` (line 1072).  This other block is only one line of code, invoking `writeBackupToURL::` (line 1069).

To verify this, I commented out the assignment to _contents in `duplicateAndReturnError:`.  Result: Now `writeToURL:::::` invokes `contentsForURL::::` which creates the big worker block that later invokes `writeAdditionalContent::::`, which I think is what we want.  To fix this, maybe the 1-line block (line 1069) needs to also call `writeAdditionalContent::::`.  Or maybe it needs more lines to be factored out from the big worker block?

---

Regarding the bigger picture:  I’ve concluded that Apple/your approach is not good for my particular application, so I've re-implemented it.

My “additional content” files are each data from the results of a scientific experiment, each typically several MB.  There may a few dozen of them, but only a few are read into memory for display.  They are rarely deleted, and _never_ edited, because, of course, that would violate scientific integrity.  My `writeAdditionalContent::::` implementation only writes one or two additional files, after a new experiment has concluded, containing the results of that new experiment.

After hacking BSManagedDocument as noted above, I spent a couple hours piling on kludges in an effort to follow the different file URLs and make it copy _all_ additional content.  Recognizing that I was getting into the NSDocument Rabbit Hole, I abandoned that approach.

The solution I've implemented for this app is to remove the _File > Duplicate_ menu item (tricky because it's not in the xib) and replaced it with a _File > Share_ menu item.  _File > Share_ shows an NSSavePanel whose accessory view asks the user which experiment results and other Core Data data from the current document should be included in a new document, then copies the document package using NSFileManager, removes the unwanted Core Data data with NSBatchDeleteRequest, and finally adds a QuickLook preview.  It doesn’t do the fancy animation of one document window _leaping_ from the original.  Opening the new document after creation is, in fact, optional.

One could argue that this app should _not_ be document-based.  I'm thinking about that.
